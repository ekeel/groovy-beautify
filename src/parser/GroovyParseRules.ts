import ParseRule from "./ParseRule";

export default [
  new ParseRule("blockcomment", { start: "/*", end: "*/", exclusive: true }),
  new ParseRule("linecomment", { start: "//", end: /(?=\n)/, exclusive: true }),
  new ParseRule("multilinestring", { start: "'''", end: "'''", exclusive: true }),
  new ParseRule("multilinestring", { start: '"""', end: '"""', exclusive: true }),
  new ParseRule("string", { start: '"', end: '"', skip: '\\"', exclusive: true }),
  new ParseRule("string", { start: "'", end: "'", skip: "\\'", exclusive: true }),
  new ParseRule("regexp", { start: /\//g, end: /\//g, skip: /\'/g, exclusive: true }),
  new ParseRule("numeric", { start: /0[xX][0-9a-fA-F]+\b/g }),
  new ParseRule("numeric", { start: /[+-]?\d+(?:(?:\.\d*)?(?:[eE][+-]?\d+)?)?\b/g }),
  new ParseRule("boolean", { start: /(?:true|false)\b/g }),
  new ParseRule("block", { start: "{", end: "}" }),
  new ParseRule("square", { start: ["?[", "["], end: "]" }),
  new ParseRule("round", { start: "(", end: ")" }),
  new ParseRule("dot", { start: /((\?|\*|\&)+)?\.(?=[^\d])(\*)?/g }), // Includes dot, spread dot, method reference, safe dot, safe chain dot and import all
  new ParseRule("delimiters", { start: [",", ":", ";"] }),
  new ParseRule("incdec", { start: ["--", "++", "**"] }),
  new ParseRule("negation", { start: "!" }),
  // prettier-ignore
  new ParseRule("operators", { start: ['--', '-=', '->', '-', '::', '!==', '!=', '?:', '?=', '?', '..<', '..', '**=', '**', 
      '*=', '*', '/=', '/', '&&', '&=', '&', '%=', '%', '^=', '^', '++', '+=', '+', '<..<', '<..', '<<=', '<<', '<=>', '<=', '<', 
      '===', '==~', '==', '=~', '=', '>=', '>>=', '>>>=', '>>', '>', '|=', '||', '|', '~'] }),
  // prettier-ignore
  new ParseRule("keywords", {
    start: ['abstract', 'assert', 'boolean', 'break', 'byte', 'case', 'catch', 'char', 'class', 'continue', 
        'default', 'def', 'double', 'do', 'else', 'enum', 'extends', 'finally', 'final', 'float', 'for', 'goto', 'if', 
        'implements', 'import', 'instanceof', 'interface', 'int', 'long', 'native', 'new', 'package', 'private', 'protected', 
        'public', 'return', 'short', 'static', 'strictfp', 'super', 'switch', 'synchronized', 'this', 'throws', 'throw', 
        'transient', 'try', 'void', 'volatile', 'while', 'with'],
    wholeword: true,
  }),
  new ParseRule("identifiers", { start: /[a-zA-Z_$][a-zA-Z0-9_$]*\b/g, wholeword: true }),
];
