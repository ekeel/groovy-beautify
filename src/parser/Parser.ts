import ParseRule, { MatchExpression } from "./ParseRule";

export type CodeBlock = {
  type: string;
  start: string;
  end?: string;
  children?: Array<CodeBlock>;
};

export class Parser {
  text: string;
  textLength: number;
  textPosition: number;
  rules: ParseRule[];

  /**
   * @param text - Text to parse
   * @param rules - Array of parsing rules. If none of specified rules match default rule with type text will be used.
   */
  constructor(text: string, rules?: ParseRule[]) {
    this.text = text;
    this.textLength = text ? text.length : 0;
    this.textPosition = 0;
    this.rules = rules || [];
  }

  scan() {
    if (this.textPosition >= this.textLength) return null;

    return this.text[this.textPosition++];
  }

  isDone() {
    return this.textPosition >= this.textLength;
  }

  move(length: number) {
    this.textPosition += length;
  }

  match(expression?: MatchExpression | MatchExpression[], wholeword?: boolean): string | undefined {
    let result;
    if (expression instanceof RegExp) {
      const modifiedExpression = new RegExp("^(" + expression.source + ")", expression.flags);
      const text = this.text.substring(this.textPosition).match(modifiedExpression)?.[0];
      if (text?.length) {
        result = text;
      }
    } else if (expression instanceof Array) {
      for (const exp of expression) {
        let match = this.match(exp);
        if (match) {
          result = match;
          break;
        }
      }
    } else if (typeof expression === "string") {
      if (this.text.substring(this.textPosition).startsWith(expression)) {
        result = expression;
      }
    }
    if (result && wholeword) {
      const prevSymbol = this.text.charAt(this.textPosition - 1);
      const nextSymbol = this.text.charAt(this.textPosition + result.length);
      if (prevSymbol.match(/[A-Za-z0-9_]/) || nextSymbol.match(/[A-Za-z0-9_]/)) {
        return undefined;
      }
    }
    return result;
  }

  matchStart(rule: ParseRule) {
    return this.match(rule.start, rule.wholeword);
  }
  matchEnd(rule: ParseRule) {
    return this.match(rule.end, rule.wholeword);
  }
  matchSkip(rule: ParseRule) {
    return this.match(rule.skip, rule.wholeword);
  }

  createDefaultTextObject(text: string) {
    return {
      type: "text",
      start: text,
    };
  }

  parse() {
    const root: CodeBlock = { type: "root", start: "" };
    const ruleStack = [] as Array<ParseRule>;
    const objStack = [root];

    while (!this.isDone()) {
      const activeRule = ruleStack[ruleStack.length - 1];
      const activeObj = objStack[objStack.length - 1];
      const activeObjLastChild = activeObj.children?.[activeObj.children.length - 1];

      if (activeRule) {
        // Check skip first
        if (activeRule.skip) {
          const activeRuleSkipText = this.matchSkip(activeRule);
          if (activeRuleSkipText) {
            // Append to existing text block or create a new one
            if (activeObjLastChild && activeObjLastChild.type === "text") {
              activeObjLastChild.start += activeRuleSkipText;
            } else {
              if (!activeObj.children) {
                activeObj.children = [];
              }
              activeObj.children.push(this.createDefaultTextObject(activeRuleSkipText));
            }
            this.move(activeRuleSkipText.length);
            continue;
          }
        }
        if (activeRule.end) {
          // Check rule end
          const activeRuleText = this.matchEnd(activeRule);
          if (activeRuleText) {
            activeObj.end = activeRuleText;
            ruleStack.pop();
            objStack.pop();
            this.move(activeRuleText.length);
            continue;
          }
        }
      }
      // Check rule start
      if (!activeRule?.exclusive) {
        const matchedRule = this.rules.find((r) => this.matchStart(r));
        if (matchedRule) {
          const matchedRuleText = this.matchStart(matchedRule);
          if (matchedRuleText) {
            const newObj = {
              type: matchedRule.name,
              start: matchedRuleText,
            };
            if (!activeObj.children) {
              activeObj.children = [];
            }
            activeObj.children.push(newObj);
            // If we need to find matched rule
            if (matchedRule.end) {
              ruleStack.push(matchedRule);
              objStack.push(newObj);
            }
            this.move(matchedRuleText.length);
            continue;
          }
        }
      }

      // Read char by char
      const ch = this.scan();
      if (ch) {
        if (activeObjLastChild && activeObjLastChild.type === "text") {
          activeObjLastChild.start += ch;
        } else {
          if (!activeObj.children) {
            activeObj.children = [];
          }
          activeObj.children.push(this.createDefaultTextObject(ch));
        }
      }
    }

    return root;
  }
}
