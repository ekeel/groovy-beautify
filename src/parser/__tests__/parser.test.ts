import GroovyParseRules from "../GroovyParseRules";
import { Parser } from "../Parser";

test("Keywords", () => {
  const parser = new Parser(`def defaultVariable = true`, GroovyParseRules);
  const result = parser.parse();
  expect(result).toMatchSnapshot();
});

test("String escaping", () => {
  const code = `"Hello \\"World\\"!"`;
  const parser = new Parser(code, GroovyParseRules);
  const parsingResult = parser.parse();
  expect(parsingResult).toMatchSnapshot();
});

export {};
