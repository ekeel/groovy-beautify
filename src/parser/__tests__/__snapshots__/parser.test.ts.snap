// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dots 1`] = `
Object {
  "children": Array [
    Object {
      "start": "
  ",
      "type": "text",
    },
    Object {
      "start": "test",
      "type": "identifiers",
    },
    Object {
      "start": ".",
      "type": "dot",
    },
    Object {
      "start": "test",
      "type": "identifiers",
    },
    Object {
      "end": ")",
      "start": "(",
      "type": "round",
    },
    Object {
      "start": " ",
      "type": "text",
    },
    Object {
      "children": Array [
        Object {
          "start": " dot",
          "type": "text",
        },
      ],
      "end": "",
      "start": "//",
      "type": "linecomment",
    },
    Object {
      "start": "
  ",
      "type": "text",
    },
    Object {
      "start": "test",
      "type": "identifiers",
    },
    Object {
      "start": "?.",
      "type": "dot",
    },
    Object {
      "start": "test",
      "type": "identifiers",
    },
    Object {
      "end": ")",
      "start": "(",
      "type": "round",
    },
    Object {
      "start": " ",
      "type": "text",
    },
    Object {
      "children": Array [
        Object {
          "start": " safe dot",
          "type": "text",
        },
      ],
      "end": "",
      "start": "//",
      "type": "linecomment",
    },
    Object {
      "start": "
  ",
      "type": "text",
    },
    Object {
      "start": "test",
      "type": "identifiers",
    },
    Object {
      "start": "??.",
      "type": "dot",
    },
    Object {
      "start": "test",
      "type": "identifiers",
    },
    Object {
      "end": ")",
      "start": "(",
      "type": "round",
    },
    Object {
      "start": " ",
      "type": "text",
    },
    Object {
      "children": Array [
        Object {
          "start": " safe chain dot",
          "type": "text",
        },
      ],
      "end": "",
      "start": "//",
      "type": "linecomment",
    },
    Object {
      "start": "
  ",
      "type": "text",
    },
    Object {
      "start": "test",
      "type": "identifiers",
    },
    Object {
      "start": "*.",
      "type": "dot",
    },
    Object {
      "start": "test",
      "type": "identifiers",
    },
    Object {
      "end": ")",
      "start": "(",
      "type": "round",
    },
    Object {
      "start": " ",
      "type": "text",
    },
    Object {
      "children": Array [
        Object {
          "start": " spread dot",
          "type": "text",
        },
      ],
      "end": "",
      "start": "//",
      "type": "linecomment",
    },
    Object {
      "start": "
  ",
      "type": "text",
    },
    Object {
      "start": "test",
      "type": "identifiers",
    },
    Object {
      "start": "&.",
      "type": "dot",
    },
    Object {
      "start": "test",
      "type": "identifiers",
    },
    Object {
      "end": ")",
      "start": "(",
      "type": "round",
    },
    Object {
      "start": " ",
      "type": "text",
    },
    Object {
      "children": Array [
        Object {
          "start": " method reference",
          "type": "text",
        },
      ],
      "end": "",
      "start": "//",
      "type": "linecomment",
    },
    Object {
      "start": "
  ",
      "type": "text",
    },
    Object {
      "start": "test",
      "type": "identifiers",
    },
    Object {
      "start": ".*",
      "type": "dot",
    },
    Object {
      "start": " ",
      "type": "text",
    },
    Object {
      "children": Array [
        Object {
          "start": " import all",
          "type": "text",
        },
      ],
      "end": "",
      "start": "//",
      "type": "linecomment",
    },
    Object {
      "start": "
  ",
      "type": "text",
    },
  ],
  "start": "",
  "type": "root",
}
`;

exports[`Keywords 1`] = `
Object {
  "children": Array [
    Object {
      "start": "def",
      "type": "keywords",
    },
    Object {
      "start": " ",
      "type": "text",
    },
    Object {
      "start": "defaultVariable",
      "type": "identifiers",
    },
    Object {
      "start": " ",
      "type": "text",
    },
    Object {
      "start": "=",
      "type": "operators",
    },
    Object {
      "start": " ",
      "type": "text",
    },
    Object {
      "start": "true",
      "type": "identifiers",
    },
  ],
  "start": "",
  "type": "root",
}
`;

exports[`String escaping 1`] = `
Object {
  "children": Array [
    Object {
      "children": Array [
        Object {
          "start": "Hello \\\\\\"World\\\\\\"!",
          "type": "text",
        },
      ],
      "end": "\\"",
      "start": "\\"",
      "type": "string",
    },
  ],
  "start": "",
  "type": "root",
}
`;
